jmclient/jmclient/wallet.py

```diff
@@ -1032,6 +999,499 @@ def close(self):
+
+        if psbt_output.derivation_map:
+            # See note to derivation map in hr_psbt_in()
+            outdict["derivation-map"] = {bintohex(k): bintohex(v.serialize(
+                )) for k, v in psbt_output.derivation_map.items()}
+
+        if psbt_output.redeem_script:
+            outdict["redeem-script"] = bintohex(psbt_output.redeem_script)
+        if psbt_output.witness_script:
+            outdict["witness-script"] = bintohex(psbt_output.witness_script)
+
+        if psbt_output.proprietary_fields:
+            outdict["proprietary-fields"] = str(psbt_output.proprietary_fields)
+        if psbt_output.unknown_fields:
+            outdict["unknown-fields"] = str(psbt_output.unknown_fields)
+
+        return outdict
+
+    @staticmethod
+    def witness_utxos_to_psbt_utxos(utxos):
+        """ Given a dict of utxos as returned from select_utxos,
+        convert them to the format required to populate PSBT inputs,
+        namely CTxOut. Note that the non-segwit case is different, there
+        you should provide an entire CMutableTransaction object instead.
+        """
+        return [btc.CMutableTxOut(v["value"],
+                                  v["script"]) for _, v in utxos.items()]
+
+    def create_psbt_from_tx(self, tx, spent_outs=None):
+        """ Given a CMutableTransaction object, which should not currently
+        contain signatures, we create and return a new PSBT object of type
+        btc.PartiallySignedTransaction.
+        Optionally the information about the spent outputs that is stored
+        in PSBT_IN_NONWITNESS_UTXO, PSBT_IN_WITNESS_UTXO and PSBT_IN_REDEEM_SCRIPT
+        can also be provided, one item per input, in the tuple (spent_outs).
+        These objects should be either CMutableTransaction, CTxOut or None,
+        Note that redeem script information cannot be provided for inputs which
+        we don't own.
+        """
+        # TODO: verify tx contains no signatures as a sanity check?
+        new_psbt = btc.PartiallySignedTransaction(unsigned_tx=tx)
+        if spent_outs is None:
+            # user has not provided input script information; psbt
+            # will not yet be usable for signing.
+            return new_psbt
+        for i, txinput in enumerate(new_psbt.inputs):
+            if spent_outs[i] is None:
+                # as above, will not be signable in this case
+                continue
+            if isinstance(spent_outs[i], (btc.CTransaction, btc.CTxOut)):
+                # note that we trust the caller to choose Tx vs TxOut as according
+                # to non-witness/witness:
+                txinput.utxo = spent_outs[i]
+            else:
+                assert False, "invalid spent output type passed into PSBT creator"
+        # we now insert redeemscripts where that is possible and necessary:
+        for i, txinput in enumerate(new_psbt.inputs):
+            if isinstance(txinput.utxo, btc.CTxOut):
```

Comment:
Another place where checking `txinput.witness_utxo` would be better after proposed change in bitcointx

Reply:
Done in  https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/9933408ad70e717014ecbb24796d104f3263ac05 . Also changed to `set_utxo` in line 1179 above.